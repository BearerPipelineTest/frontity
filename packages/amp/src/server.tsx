import { CacheProvider } from "frontity";
import createEmotionServer from "@emotion/server/create-instance";
import createCache from "@emotion/cache";
import config, { CACHE_KEY } from "./config";
import AMP from "../types";

export default {
  ...config,
  actions: {
    amp: {
      ...config.actions.amp,
      beforeSSR({ libraries }) {
        // Grab a copy of the previous render method.
        const render = libraries.frontity.render;

        // Define the emotion css extraction one.
        libraries.frontity.render = ({ App, ...rest }) => {
          const cache = createCache({ key: CACHE_KEY });
          const { extractCritical } = createEmotionServer(cache);

          const result = extractCritical(
            render({
              ...rest,
              App: function Wrapped() {
                return (
                  <CacheProvider value={cache}>
                    <App />
                  </CacheProvider>
                );
              },
            })
          );

          return result;
        };

        // Grab a copy of the previos template method.
        const template = libraries.frontity.template;

        // Define the emotion style tag with the render result.
        libraries.frontity.template = ({ result, ...rest }) => {
          const { html, ids, css } = result;

          // Push the emotion style tag.
          rest.head.push(
            `<style data-emotion="${CACHE_KEY} ${ids.join(" ")}">${css}</style>`
          );

          return template({
            ...rest,
            html,
          });
        };
      },
    },
  },
} as AMP;
