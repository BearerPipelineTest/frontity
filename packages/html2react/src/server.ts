import Html2React from "../types";
import Component from "./libraries/component";
import parse from "./libraries/parse/himalaya-parser";
import script from "../processors/script";

const html2react: Html2React = {
  name: "@frontity/html2react",
  libraries: {
    html2react: {
      processors: [script],
      Component,
      parse,
    },
  },
};

export default html2react;
