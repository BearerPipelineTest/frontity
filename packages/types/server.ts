import { Middleware, Next, ParameterizedContext } from "koa";
import { ResolveState, ResolveActions } from "./utils";
import Package from "./package";

/**
 * Frontity context for Koa middleware.
 */
export interface FrontityContext<Packages extends Package>
  extends ParameterizedContext {
  /**
   * Frontity namespace for custom properties.
   */
  frontity: any;

  /**
   * Packages state.
   */
  state: ResolveState<Packages["state"]>;

  /**
   * Packages actions.
   */
  actions: ResolveActions<Packages["actions"]>;

  /**
   * Packages libraries.
   */
  libraries: Packages["libraries"];

  /**
   * Packages server middleware.
   */
  server: Packages["server"];

  /**
   * Extended Koa context.
   */
  ctx: FrontityContext<Packages>;

  /**
   * Koa Next function.
   */
  next: Next;
}

/**
 * Types for server middleware in Frontity packages.
 */
export type Server<Packages extends Package> = Middleware<
  ResolveState<Packages["state"]>,
  FrontityContext<Packages>
>;

/**
 * Types for async server middleware in Frontity packages.
 */
export type AsyncServer<Packages extends Package> = Server<Packages>;
