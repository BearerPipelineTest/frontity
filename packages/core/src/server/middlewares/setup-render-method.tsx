import { Middleware, Next } from "koa";
import { Context } from "@frontity/types";
import { renderToStaticMarkup, renderToString } from "react-dom/server";

/**
 * The default serializer. Handles the case where there are no entrypoints.
 *
 * @param args - The arguments to be used for this pass.
 *
 * @returns The function to serialize with.
 */
const serialize = ({ collectChunks, hasEntryPoint }: any) => (jsx: any) => {
  if (!hasEntryPoint) {
    return renderToStaticMarkup(jsx);
  }

  return renderToString(collectChunks(jsx));
};

/**
 * The render method for a given React App.
 *
 * @param args - The arguments to be used for render.
 *
 * @returns The serialized App.
 */
const render = ({ App, ...args }) => {
  return serialize(args)(<App />);
};

/**
 * Defines the module stats for the current request.
 *
 * @param ctx - Koa context.
 * @param next - The next method.
 *
 * @returns The awaited next.
 */
export const setupRenderMethod = async (
  ctx: Context,
  next: Next
): Promise<Middleware> => {
  const { store } = ctx.state;

  // Get the render method.
  ctx.state.render = render;

  // If the user has defined a custom render, used that and pass
  // along the default frontity renderer.
  if (store.libraries.frontity.render) {
    ctx.state.render = (args: any) => {
      return store.libraries.frontity.render({
        ...args,
        defaultRenderer: serialize(args),
      });
    };
  }

  return await next();
};
