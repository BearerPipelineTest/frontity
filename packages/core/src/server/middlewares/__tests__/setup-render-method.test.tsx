import { FC } from "react";
import { renderToStaticMarkup, renderToString } from "react-dom/server";
import { setupRenderMethod } from "../setup-render-method";
import { createKoaContext } from "./__utilities__/create-koa-context";
import { fakeNext } from "./__utilities__/fake-next";

jest.mock("react-dom/server", () => ({
  renderToString: jest.fn(),
  renderToStaticMarkup: jest.fn(),
}));

describe("setupRenderMethod", () => {
  it("should define a default render", async () => {
    const ctx: any = createKoaContext();

    await setupRenderMethod(ctx, fakeNext);

    expect(ctx.state.render).toBeDefined();
  });

  it("should call the collect chunks if we have entry points", async () => {
    const ctx: any = createKoaContext();

    await setupRenderMethod(ctx, fakeNext);

    const collectChunks = jest.fn();
    const App: FC = () => <div>app</div>;

    ctx.state.render({ collectChunks, App });

    // Since we render without the entry point flag
    // the collectChunks should not be called.
    expect(collectChunks).not.toBeCalled();

    // Call the render with the hasEntryPoints flag
    ctx.state.render({
      collectChunks,
      App,
      hasEntryPoint: true,
    });

    expect(collectChunks).toBeCalled();
  });

  describe("custom render", () => {
    it("should wrap the user defined render", async () => {
      function customRender() {
        return "custom";
      }

      const ctx: any = createKoaContext({
        render: customRender,
      });

      await setupRenderMethod(ctx, fakeNext);

      expect(ctx.state.render({})).toEqual("custom");
    });

    it("should call the render with a defaultRenderer", async () => {
      const customRenderMethod = jest.fn();
      const ctx: any = createKoaContext({
        render: customRenderMethod,
      });

      await setupRenderMethod(ctx, fakeNext);

      const collectChunks = jest.fn();
      const App: FC = () => <div>custom</div>;

      ctx.state.render({
        collectChunks,
        App,
        hasEntryPoint: true,
      });

      // Expect the custom render method to be called.
      expect(customRenderMethod).toBeCalled();

      // Expect the collect chunks not to be called
      // since the custom render method is mocked.
      expect(collectChunks).not.toBeCalled();

      // Expect the arguments to be consistent.
      expect(
        Object.keys(customRenderMethod.mock.calls[0][0])
      ).toMatchSnapshot();
    });

    it("defaultRenderer should serialize accordingly", async () => {
      const ctx: any = createKoaContext({
        render: ({ App, defaultRenderer }) => {
          return defaultRenderer(<App />);
        },
      });

      await setupRenderMethod(ctx, fakeNext);

      const collectChunks = jest.fn();
      const App: FC = () => <div>custom</div>;

      ctx.state.render({
        collectChunks,
        App,
        hasEntryPoint: true,
      });

      // Expect the collect chunks not to be called
      // since the custom render method is mocked.
      expect(collectChunks).toBeCalledTimes(1);

      // Expec the render to string to be called.
      expect(renderToString).toBeCalled();

      collectChunks.mockReset();

      // Render without an entry point.
      ctx.state.render({
        collectChunks,
        App,
      });

      // Expect the collect chunks not to be called
      // since the custom render method is mocked.
      expect(collectChunks).not.toBeCalled();

      // Expec the render to static markup to be called.
      expect(renderToStaticMarkup).toBeCalled();
    });
  });
});
