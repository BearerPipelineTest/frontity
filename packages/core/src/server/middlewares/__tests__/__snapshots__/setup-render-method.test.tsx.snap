// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupRenderMethod custom render should call the render with a defaultRenderer 1`] = `
Array [
  "collectChunks",
  "App",
  "hasEntryPoint",
  "defaultRenderer",
]
`;

exports[`setupRenderMethod custom render should wrap the user defined render 1`] = `
"(args) => {
            return store.libraries.frontity.render(Object.assign(Object.assign({}, args), { defaultRenderer: serialize(args) }));
        }"
`;

exports[`setupRenderMethod should define a default render 1`] = `
"(_a) => {
    var { App } = _a, args = __rest(_a, [\\"App\\"]);
    return serialize(args)(jsx_runtime_1.jsx(App, {}, void 0));
}"
`;
